<#13> feat : User 기능 구현

nrudev opened the pull request:
## 내용
아래 기능 구현 완료
- 로그인한 사용자 본인 정보 조회
- 회원정보 수정
- 회원탈퇴
- 닉네임 중복 조회
- 사용자 접근 권한 조회

## 참고
- API 명세서에 적힌 '작성한 게시물 목록 조회', '좋아요한 게시물 목록 조회'는 위 기능들이 먼저 develop에 merge된 후 구현할 예정

nrudev committed:
feat : 로그인한 사용자 본인 정보 조회

nrudev committed:
refactor : UserDto 이름 및 내용 수정

nrudev committed:
feat : 회원정보 수정 기능 구현 및 코드 리팩토링

- 회원정보 수정 기능 구현
- Controller 에서 Session user = (SessionUser) httpSession.getAttribute("user"); 가 반복되어 어노테이션으로 만들어서 파라미터로 넣는 것으로 변경
- Service에서 회원조회 유효성 검사가 반복되어 findUser() 메서드로 만들어서 코드 간결화
- Response DTO 에 User를 파라미터로 받는 생성자를 만들어 엔티티 클래스를 DTO 로 만들때 코드 간결화

nrudev committed:
feat : 회원탈퇴 기능 구현

nrudev committed:
feat : 닉네임 중복 조회 기능 구현

nrudev committed:
feat : 사용자 접근 권한 조회 기능 구현

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/config/auth/SecurityConfig.java:
```diff
@@ -18,7 +18,7 @@ protected void configure(HttpSecurity http) throws Exception {
                 .and()
                 .authorizeRequests()
-                .anyRequest().permitAll()
+                .anyRequest().authenticated()
```
요건 추후에 추가하는 건 어떨까요? 단순 열람 시에도 권한 제한이 걸리면 프론트님이 번거로우실 것 같아서요 😅  아니면 개발 기간동안 세션 유지 시간을 어마무시하게 늘려도 좋을 것 같아요!

nrudev replied:
테스트해보려고 바꿔놓은 거에요! 돌려놓겠습니다~

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+
+    private final UserService userService;
+
+    @GetMapping("/user-info")
+    public ResponseEntity<UserDetailResponse> getUserDetail(@LoginUser SessionUser user) {
+
+        if (user == null) {
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
+        }
+
+        return new ResponseEntity<>(userService.getUserDetail(user), HttpStatus.OK);
+    }
```
마이페이지 api가 맞을까요? 본인만 열람 가능하면 이를 좀더 돋보일 수 있는 네이밍을 사용하고, 다른 유저들도 열람할 수 있는 거라면 `/api/users/{userId}`도 괜찮지 않을까 싶습니다!

dadahee replied:
지금은 모든 request에 대해 로그인 권한을 필요로 하도록 설정해 두었으니(위의 `SecurityConfig` 참조) 유저의 null 체킹은 하지 않아도 될듯합니다. !!

nrudev replied:
마이페이지에서 내정보 조회입니다! 다른 사람들이 다른 회원 개인정보를 조회하면 안되니까요.
관리자가 열람하는 건 admin에서 별도로 하는 편이 좋을 듯 합니다.
permitAll()로 바꿔놓을테니 유효성검사는 유지하겠습니다~

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+    @PutMapping ("/users/{userId}")
+    public ResponseEntity<UserResponse> updateUserInfo(@PathVariable Long userId, @LoginUser SessionUser user, @RequestBody UpdateUserRequest request) {
+
+        if (user == null || user.getUserId() != userId) {
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
+        }
```
null 체킹은 위에서 언급한대로이고, user id 일치 여부와 같은 로직은 서비스 레이어에서 처리한다고 알고 있는데 혹시 제가 잘못 알고 있는 걸까요?

nrudev replied:
null 체킹 한번 하면서, 동시에 user id 일치 여부를 컨트롤러에서 먼저 확인하는 편이 좋을 것이라 생각했는데, 서비스 레이어에서 처리하는 것과 어떤 차이가 있는지 알 수 있을까요?

dadahee replied:
컨트롤러는 요청과 요청에 대한 응답 & 처리 로직을 매핑해주는 레이어이기 때문에, 스프링 내부에서 제공하는 어노테이션 & 함수 외의 서비스 로직, 유효성 검사 등은 모두 컨트롤러에서 가져가면 안 된다고 생각했습니다 !

nrudev replied:
맞는 말씀이네요.
근데 지금 보니 서비스 레이어로 옮기게 되면 ResponseEntity 로 반환하기 어려울 것 같은데, 어떻게 하면 좋을까요?

dadahee replied:
원래 인터뷰 구현까지 끝나고 예외처리쪽 건드릴 생각이었는데, 전역 exception handler를 추가하고, 서비스 안에서 예외를 발생시키면 될 거예요! 일단 예외 부분은 표시만 해두고 설계해도 될 거 같습니다ㅎㅎ

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
+        }
+
+        return new ResponseEntity<>(userService.update(user, request), HttpStatus.OK);
```
ResponseEntity.ok(`Response 객체`)를 사용하면 훨씬 간결해져서, 이 스타일로 통일하는 건 어떠실까요!

nrudev replied:
좋습니다!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+    @PostMapping("/users/validate-nickname")
+    public ResponseEntity<Boolean> validateNickname(@LoginUser SessionUser user, @RequestBody ValidateNicknameRequest request) {
+
+        if (user == null) {
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
+        }
+        if (request.equals(user.getNickname())) {
+            return new ResponseEntity<>(false, HttpStatus.OK);
+        }
+
+        // true = 닉네임 중복. 사용 불가.
+        // false = 닉네임 사용 가능.
```
사용 가능한 닉네임인지 보려면 유저 레포지토리에서 중복되는 닉네임이 있는지 조회해봐야 하지 않나요?
또, 간단한 로직이지만 서비스 레이어 안에 들어갈 수 있도록 구성하고, Response DTO를 두어 boolean 값과 닉네임을 모두 리턴해주는 것이 api 요청/응답 결과를 한 눈에 파악하기 쉽지 않나 싶습니다! 어떻게 생각하시나요?

nrudev replied:
중복 닉네임 조회는 서비스 레이어 확인하시면 될 것 같습니다!
response dto 하나 추가하겠습니다~

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+    }
+
+    @PatchMapping("/users/{userId}")
+    public ResponseEntity<Long> deleteUser(@PathVariable Long userId, @LoginUser SessionUser user) {
```
저는 Void 타입으로 리턴하여... 저희끼리 통일이 필요할 것 같습니다...

dadahee replied:
일단 저는 생성: 201, 수정: 200, 삭제: 204로 설계하였는데, 희승님께서 구성하신 방식과 이유가 궁금해요.
저도 많이 부족해서 여쭤보는 겁니다...ㅎㅎ;;

nrudev replied:
그럼 저도 Void로 통일하고, 반환을 204로 하는 것으로 수정하겠습니다!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+        if (user == null || user.getUserId() != userId) {
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
+        }
+        return new ResponseEntity<>(userService.withdraw(user), HttpStatus.OK);
```
해당 내용은 앞에서도 언급하였으므로 앞으로 생략하겠습니다 !

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+    public ResponseEntity<Boolean> validateNickname(@LoginUser SessionUser user, @RequestBody ValidateNicknameRequest request) {
+
+        if (user == null) {
+            return new ResponseEntity<>(HttpStatus.FORBIDDEN);
```
아 그리고 추가적으로 말씀드리면 유저가 Null인 경우는 로그인하지 않은 경우이므로 401 응답 코드가 더 적절해 보입니다.
[401 vs 403 응답코드](https://mangkyu.tistory.com/146) 참고하시면 도움이 될 것 같습니다 !

nrudev replied:
401이 더 적합해보이네요. 401로 변경할게요~

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/request/UpdateUserRequest.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@NoArgsConstructor
+@AllArgsConstructor
```
Request 클래스 생성자를 사용할 일이 있을까요..? 더미 데이터 주입용으로 추가해두신 건가요!

nrudev replied:
어제 제가 작성한 글 1번 참고해주시면 감사하겠습니다:)
https://nrudev.github.io/til/211206/

nrudev replied:
@RequestBody 를 받는 과정에서 JSON이 DTO 로 역직렬화가 되어야 하는데, 기본 생성자가 없으면 역직렬화가 되지 않습니다. 그래서 @NoArgsConstructor 를 추가해준 거에요.

dadahee replied:
자세한 설명 감사합니다😁 역직렬화에 `@AllArgsConstructor` 내용은 해당이 안 되는 걸로 알고 있는데, 얘는 빼도 되지 않을까요?

nrudev replied:
아ㅋㅋㅋ request dto 에 변수가 nickname 뿐이라서 생성자를 롬복으로 대체한건데 그냥 이전처럼 생성자 직접 작성해서 넣어줄까요?? 어느 쪽이든 결과가 똑같긴 합니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/request/UpdateUserRequest.java:
```diff
@@ -0,0 +1,15 @@
+public class UpdateUserRequest {
+
+//    private String password;
+    private String nickname;
```
validation 관련 어노테이션을 추가해주시면 좋을 것 같습니다 :)

nrudev replied:
아! validation 추가하겠습니다ㅎㅎ

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/request/ValidateNicknameRequest.java:
```diff
@@ -0,0 +1,14 @@
+
+    private String nickname;
+
+}
```
위에서 언급한 내용과 동일합니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/response/UserDetailResponse.java:
```diff
@@ -0,0 +1,40 @@
+        this.role = role;
+        this.provider = provider;
+        this.createdAt = createdAt;
+    }
```
몰라서 여쭤봅니다만, 클래스 전체에 어노테이션을 붙이는 것과 해당 부분에만 붙이는 것은 무슨 차이가 있나요?
제가 이해하기로는 전체 생성자에 직접 주입하면 전체 필드를 빌드할 때만 빌더를 이용 가능한 것으로 알고 있는데, 맞나요?

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/response/UserDetailResponse.java:
```diff
@@ -0,0 +1,40 @@
+    private String nickname;
+    private Role role;
+    private Provider provider;
+    private LocalDateTime createdAt;
```
변경될 일이 없으니(큰일나니) final 선언을 해주면 좋을 것 같습니다.

dadahee replied:
더불어 저는 createdAt에는 공통적으로 `@JsonFormat` 어노테이션을 붙여주었습니다. 시간 필드의 경우 다양한 어노테이션이 있고, 이유도 다양하니, 검색해 보신 후에 이야기 나누면 좋을 것 같습니다!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/response/UserResponse.java:
```diff
@@ -0,0 +1,27 @@
+
+    private Long id;
+    private String email;
+    private String nickname;
```
마찬가지로 final 선언을 해주시면 좋을 것 같습니다. 또한 저는 interview, tag, comment 등 다양한 도메인에서 간단한 유저 정보(id, nickname)를 리턴할 필요가 있어 글로벌 dto로 이동하였습니다! 해당 dto와 네이밍이 겹치네요. `UpdateUserResponse` 등으로 수정함은 어떨까 싶습니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/response/UserRoleResponse.java:
```diff
@@ -0,0 +1,15 @@
+import lombok.NoArgsConstructor;
+
+@AllArgsConstructor
+@NoArgsConstructor
```
개인적인 견해이지만 어노테이션은 최소한만 붙이는 게 좋다고 생각합니다. 두 생성자 어노테이션을 사용한 이유가 있으신지 궁금해요.

nrudev replied:
어떤 어노테이션인지는 아실 거라 생각합니다.
제가 아는 바로는 파라미터를 받는 생성자가 있을 경우, 별도로 기본 생성자를 작성해주는 편이 좋다고 알고 있습니다.
굳이 아래에 생성자를 쓸 필요 없이 lombok 으로 간편하게 하는 편이 좋을 것 같아 그렇게 처리했습니다.

dadahee replied:
제가 리뷰를 너무 세세하게 남기는 것 같다는 생각이 드네요..ㅎㅎ;;
생성자가 필요한 곳이 있다면 당연히 lombok을 사용하는 편이 낫겠지요! 좋습니다😁

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/domain/User.java:
```diff
@@ -59,6 +59,10 @@ public User update(String nickname) {
         return this;
     }
 
+    public void delete() {
+        this.deleted = true;
+    }
```
반대로 deleted를 false로 바꾸는 경우도 필요하지 않을까요? (수정 경로가 한정적이긴 하지만요...)
enable 함수를 만들기에는 코드가 조잡해지니 매개변수화함이 어떨까 싶습니다.

nrudev replied:
음 이 부분은 논의가 필요할 것 같습니다.
저는 이게 회원탈퇴 기능이다보니 함부로 회원을 복구시키는 기능이 바람직하지는 않다고 봐서요.

dadahee replied:
어드민에서 회원 복구도 막자는 말씀으로 이해하면 될까요?

nrudev replied:
아 네 회원복구도 고려해봄직 하지만, 필요성이 크게 느껴지진 않습니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/repository/UserRepository.java:
```diff
@@ -2,10 +2,17 @@
 
     Optional<User> findByEmail(String email);
+
+    @Query(nativeQuery = true, value = "SELECT * FROM USER WHERE id = :id AND deleted <> true")
+    Optional<User> findById(@Param("id") Long id);
```
nativeQuery 옵션과 @Param을 사용하는 건 처음 봐서 (눈물찔끔) 공부좀 하고 다시 답글 달겠습니다...

nrudev replied:
JPA 에서는 JPQL이라는 언어를 사용합니다. @Query 어노테이션은 아실거고, 해당 어노테이션의 파라미터로 주는 nativeQuery는 JPQL과 SQL을 구분해주는 역할을 해요. nativeQuery를 true로 설정해야 쿼리문에서 * 과 같은 문자를 사용할 수가 있더라구요.
아래 참고하시면 더 좋을 것 같습니다.
https://jforj.tistory.com/90

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/service/UserService.java:
```diff
@@ -0,0 +1,22 @@
+import com.j2kb5th.chippo.user.controller.dto.response.UserRoleResponse;
+import com.j2kb5th.chippo.user.domain.Role;
+
+public interface UserService {
```
어억 ㅠㅠ 저희 이 부분은 회의를 해야 할 것 같습니다.
저는 서비스 -> 컨트롤러로 데이터 전달 시 DTO가 아니라 엔티티를 주었는데, 희승님께서는 dto를 사용하신 이유가 궁금합니다.
만약 서비스 <-- dto --> 컨트롤러로 데이터를 전달하게 되면 프로젝트 패키지 구조를 변경해야 할 것 같습니다. !
현재는 제 기준(...)에 맞춰 작성하여, controller 안에 dto 패키지가 있거든요! 껄껄...

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    private Optional<User> findUser(Long userId) {
+        Optional<User> userFoundById = userRepository.findById(userId);
+
+        if (!userFoundById.isPresent()) {
+            throw new ResponseStatusException(HttpStatus.NOT_ACCEPTABLE, "탈퇴했거나 가입하지 않은 회원입니다.");
+        }
+
+        return userFoundById;
```
Optional 객체를 사용하기 때문에 `orElseThrow` 메서드로 간략하게 쓸 수 있는 걸로 알고 있습니다 !

dadahee replied:
다른 함수들 구현 부분은 해당 부분 논의가 끝난 후에 다시 리뷰를 해야 할 것 같습니다 @.@
(orElseThrow로 객체를 처리하면 바로 Entity 클래스로 객체를 떼올 수 있기 때문에 map, get등의 처리 필요 없이 사용할 수 있는 것으로 알고 있거든요...)

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+        return findUser(userId).map(user -> new UserRoleResponse(user.getRole())).get();
+    }
+
+    private Optional<User> findUser(Long userId) {
```
Optional로 바로 리턴하시는 이유가 궁금하네요. 이럴 경우 위에처럼 map을 사용하여야 하지 않나요?

nrudev replied:
위에 UserRoleResponse 로 변환하는 것을 보시면 아시겠지만 경우에 따라서 조회한 것을 바로 map 을 사용해 response dto 타입으로 변환하는 경우가 있는지라 Optional 객체로 반환하는 방식을 택했습니다.
바로 Entity 클래스 객체를 받아오면 .get() 을 줄일 수는 있지만, dto 타입으로 변환할 때 다시 Optional 로 감싸주거나 컬렉션 객체로 변환해야 해서요. UserServiceImpl 에서 update 부분이 참고가 될 것 같습니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/service/UserService.java:
```diff
@@ -0,0 +1,22 @@
+
+    UserDetailResponse getUserDetail(SessionUser user);
+
+    UserResponse update(SessionUser user, UpdateUserRequest request);
+
+    Long withdraw(SessionUser user);
```
UserService이기 때문에 어떤 내용에 관한 건지 뻔히 보이긴 한다만.. 어떤 정보를 지칭하는지 알기 편하게 `updateUser`, `withdrawUser` 등으로 수정하는 건 어떨까요?

dadahee replied:
사실 update만 보면 updateNickname일수도 있고, updateRole일 수도 있으니, 큰 단위로 User라고 지칭하면 좋지 않을까 생각이 드네요. ㅎㅎ

nrudev replied:
다희님 말씀대로 메서드명 수정하겠습니다! 그 편이 더 분명할 것 같네요ㅎㅎ
update는 추후 password를 추가할 것을 염두에 두어 다희님 말씀대로 updateUser로 하는 것이 맞다고 생각해요.

참고 : role 변경은 admin에서 구현해야 하는 기능이기 때문에 user에는 추가하지 않겠습니다!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+    private final UserRepository userRepository;
+
+    @Override
+    public UserDetailResponse getUserDetail(SessionUser user) {
```
`@Transactional(readonly = true)`를 추가해주시면 좋을 것 같습니다~ 여차저차한 이유로 readonly 체크 시에는 성능이 훨 빠르다고 하네요.

nrudev replied:
앗 몰랐네요ㅎㅎ 말씀해주신대로 추가하고, 저도 검색해서 공부해보겠습니다.

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/service/UserServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+
+    @Transactional
+    @Override
+    public Long withdraw(SessionUser user) {
```
사실상 delete, remove와 같은 맥락에서 쓰이는 걸로 알고 있습니다. 현재는 필드값만 바꿔주는 거라 네이밍을 조금 더 분명하게 수정해보면 어떨까요?

nrudev replied:
db에서 삭제하는 것이 아니기도 하고, controller 와 service 간에 메서드명에 조금 차이를 두고 싶어서 withdraw로 했는데, 차라리 delete로 하는 편이 나을까요?

dadahee reviewed (changes requested):
역시 구현부 들어가니 개개인 코드 스타일이 반영되어 얘기 나눌 내용이 많아지네요 ㅎㅎ...
리뷰 읽는 것도 피곤하실 것 같아요... 고생 많으셨습니다 !!

nrudev committed:
refactor : PR 리뷰 반영 수정 (1)

## 수정한 내용
- request dto 에 validation 추가
- response dto 변수에 final 추가
- response dto LocalDateTime 타입 변수에 `@JsonFormat` 추가
- ValidateNicknameResponse 추가
- controller 조회 기능에 `@Transactional(readOnly = true)` 추가
- controller 반환 타입 ResponseEntity.ok() 로 통일
- service 메서드명에 User 추가하여 의미 명확히
- UserServiceImpl findUser 메서드 리팩토링

## 미흡한 부분
- 회원정보 수정 후 세션값도 변경해주어야 함
- 회원탈퇴 후 로그아웃 처리해야 함

dadahee reviewed (commented):
정말 고생 많으셨습니다 ! 다음 회의 때에는 컨트롤러, 서비스간 통신 단위(엔티티vsDTO)만 정하면 될 것 같네요. 스파게티 코드가 될 수도 있겠지만... 일단은 지금까지 머지된 방식으로 구현하고 한 쪽으로 리팩토링 하는 게 어떨까 싶습니다. 구현 속도도 유지할겸, 부딪쳐보고 정할겸 목요일까지 각자 구현해보고 이야기 나눠요 !

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/request/UpdateUserRequest.java:
```diff
@@ -4,12 +4,14 @@
 public class UpdateUserRequest {
 
-//    private String password;
+    @NotBlank
```
`@Size(min = 값, max=값)`까지만 추가해주시면 완벽할 것 같네요 !!

dadahee commented on back-end/src/main/java/com/j2kb5th/chippo/user/controller/dto/response/UserUpdateResponse.java:
```diff
@@ -0,0 +1,23 @@
+@AllArgsConstructor
+@Builder
+@Getter
+public class UserUpdateResponse {
```
Request dto 네이밍에 맞춰 `UpdateUserResponse`는 어떨까요?

dadahee reviewed (approved):
추가 수정 내용 전부 확인했습니다 :) 메시지 내용 확인하시고, 추후 리뷰 없이 반영 내용 바로 머지해주시면 될 것 같습니다 👍

nrudev committed:
refactor : PR 리뷰 반영 수정 (2)

## 수정사항
- request dto validation에 @Size 추가
- UserUpdateResponse를 UpdateUserResponse 로 변경
